tests/garbage_collector_test.py

```diff
@@ -0,0 +1,66 @@
+"""Tests for aws/garbage_collector."""
+
+import os
+import unittest
+from mock import MagicMock as Mock
+from treadmill_aws import garbage_collector
+
+
+# pylint: disable=protected-access
+class GarbageCollector(unittest.TestCase):
+    """Tests garbage_collector"""
+
+    def setUp(self):
+        """Set up mocks"""
+        self.ec2_instances = {
+            'Reservations': [
+                {'Instances': [
+                    {'Tags': [{"Key": "Name",
+                               "Value": 'valid-instance.foo.bar.com'}]}]}]}
+        self.valid_instances = {"valid-instance.foo.bar.com"}
+        self.ldap_servers = {"valid-instance.foo.bar.com",
+                             "terminated-instance.foo.bar.com"}
+        self.ipa_servers = self.ldap_servers
+
+        self.admin_srv = Mock()
+        self.admin_srv.admin = Mock()
+        self.admin_srv.admin.search = Mock()
+        self.admin_srv.delete = Mock()
+
+    def tearDown(self):
+        """Tear down"""
+        os.environ["IPA_DOMAIN"] = ""
+        os.environ["IPA_CACERT"] = ""
+
+    def test_valid_ec2_instances(self):
+        """Testing jmespath query returns a list of hostnames"""
+        ec2_conn = Mock()
+        ec2_conn.describe_instances = Mock()
+        ec2_conn.describe_instances.return_value = self.ec2_instances
+        valid_instances = garbage_collector._INSTANCE_NAME.search(
+            self.ec2_instances)
+        self.assertListEqual(valid_instances, ['valid-instance.foo.bar.com'])
+
+    def test_ipa_client(self):
+        """Test that returns an IPAClient instance"""
+        pass
+
+    def test_ldap(self):
+        """Testing LDAP plugin would delete terminated instances"""
+        for hostname in self.ldap_servers - self.valid_instances:
+            self.admin_srv.delete(hostname)
+        self.admin_srv.delete.assert_called_once_with(
+            'terminated-instance.foo.bar.com')
+
+    def test_ipa(self):
+        """Testing IPA plugin would delete terminated instances"""
+        ipa_client = Mock()
+        ipa_client.unenroll_host = Mock()
+        for hostname in self.ipa_servers - self.valid_instances:
```

Comment:
same command as for ldap test.

proper test is - mock return of describe, as it looks - complex data structure. mock return of adminsrv - whatever it returns. run gc - make sure that it invokes things correctly.

other missing tests:

 - what happens if you unenroll twice. what happens if you remove from ipa twice. what happens if remove from ipa fail - these are important tests.

Reply:
I have removed tests because testing what happens if I remove a host twice from ipa or ldap is out of the scope of this module, since I did not write/modify the ipa and ldap clients.